src/main/java/mezz/jei/transfer/BasicRecipeTransferHandlerServer.java

```diff
@@ -1,281 +1,358 @@
 				continue;
 			}
 			if (craftingSlot.hasItem()) {
 				ItemStack craftingItem = craftingSlot.remove(Integer.MAX_VALUE);
 				clearedCraftingItems.add(craftingItem);
 			}
-			if (requireCompleteSets) {
-				ItemStack transferItem = toTransfer.get(craftingSlotNumberIndex);
-				if (transferItem != null) {
-					int slotStackLimit = craftingSlot.getMaxStackSize(transferItem);
-					minSlotStackLimit = Math.min(slotStackLimit, minSlotStackLimit);
-				}
-			}
 		}
+		return clearedCraftingItems;
+	}
 
-		// put items into the crafting grid
-		for (Map.Entry<Integer, ItemStack> entry : toTransfer.entrySet()) {
-			Integer craftNumber = entry.getKey();
-			Integer slotNumber = craftingSlots.get(craftNumber);
-			Slot slot = container.getSlot(slotNumber);
-
-			ItemStack stack = entry.getValue();
-			if (slot.mayPlace(stack)) {
-				if (stack.getCount() > minSlotStackLimit) {
-					ItemStack remainder = stack.split(stack.getCount() - minSlotStackLimit);
-					clearedCraftingItems.add(remainder);
+	private static List<ItemStack> putItemsIntoCraftingGrid(
+		Map<Slot, ItemStack> recipeSlotToTakenStacks,
+		boolean requireCompleteSets
+	) {
+		final int slotStackLimit = getSlotStackLimit(recipeSlotToTakenStacks, requireCompleteSets);
+		List<ItemStack> remainderItems = new ArrayList<>();
+
+		recipeSlotToTakenStacks.forEach((slot, stack) -> {
+			if (slot.getItem().isEmpty() && slot.mayPlace(stack)) {
+				ItemStack remainder = slot.safeInsert(stack, slotStackLimit);
+				if (!remainder.isEmpty()) {
+					remainderItems.add(remainder);
 				}
-				slot.set(stack);
 			} else {
-				clearedCraftingItems.add(stack);
+				remainderItems.add(stack);
 			}
-		}
+		});
 
-		// put cleared items back into the inventory
-		for (ItemStack oldCraftingItem : clearedCraftingItems) {
-			int added = addStack(container, inventorySlots, oldCraftingItem);
-			if (added < oldCraftingItem.getCount()) {
-				if (!player.getInventory().add(oldCraftingItem)) {
-					player.drop(oldCraftingItem, false);
-				}
+		return remainderItems;
+	}
+
+	@Nullable
+	private static Map<Slot, ItemStack> calculateRequiredStacks(List<Tuple<Slot, Slot>> recipeSlotToSourceSlots, Player player) {
+		Map<Slot, ItemStack> recipeSlotToRequired = new HashMap<>(recipeSlotToSourceSlots.size());
+		for (Tuple<Slot, Slot> entry : recipeSlotToSourceSlots) {
+			Slot recipeSlot = entry.getA();
+			Slot inventorySlot = entry.getB();
+			if (!inventorySlot.mayPickup(player)) {
+				LOGGER.error(
+					"Tried to transfer recipe but was given an" +
+						" inventory slot that the player can't pickup from: " + inventorySlot.index);
+				return null;
+			}
+			final ItemStack slotStack = inventorySlot.getItem();
+			if (slotStack.isEmpty()) {
+				LOGGER.error(
+					"Tried to transfer recipe but was given an" +
+						" empty inventory slot as an ingredient source: " + inventorySlot.index);
+				return null;
 			}
+			ItemStack stack = slotStack.copy();
+			stack.setCount(1);
```

Comment:
```java
ItemStack stack = ItemHandlerHelper.copyStackWithSize(slotStack, 1);
```

Reply:
It would look cleaner, but I don't see much use in relying on Forge in this specific case. 
Forge methods are moved or removed pretty frequently, and this one is not overridable or hooked into so it's not adding much benefit.

Reply:
That is fair enough, mostly wanted to point out that method exists.

Reply:
Thanks!